### PR TITLE
issue: 1438405 Update ip header after dst entry creation

### PR DESCRIPTION
**Pull request by rafiw**

### CODE REVIEW

File: src/vma/proto/dst_entry.cpp [hunk 1]
```diff
@@ -376,7 +376,9 @@ bool dst_entry::conf_l2_hdr_and_snd_wqe_eth()
 		if (src && dst) {
 		BULLSEYE_EXCLUDE_BLOCK_END
 			if (netdevice_eth->get_vlan()) { //vlan interface
-				uint16_t vlan_tci = (m_pcp << 12) | netdevice_eth->get_vlan();
+				uint16_t vlan_tci =
+						((uint16_t)m_pcp << NET_ETH_VLAN_PCP_OFFSET) |
```
Comment (liranoz12) [inline src/vma/proto/dst_entry.cpp]: pcp value was 12 before this change, are we sure about it ?

File: src/vma/proto/header.h [hunk 2]
```diff
@@ -54,7 +54,7 @@
 #define NET_IB_IP_ALIGN_SZ		16
 #define NET_ETH_IP_ALIGN_SZ		 6
 #define NET_ETH_VLAN_IP_ALIGN_SZ	 2
-
+#define NET_ETH_VLAN_PCP_OFFSET	13
```
Comment (liranoz12) [inline src/vma/proto/header.h]: Can we use kernel definition instead ?

File: src/vma/proto/dst_entry.h [hunk 3]
```diff
@@ -151,8 +153,8 @@ class dst_entry : public cache_observer, public tostr, public neigh_observer
 	virtual uint8_t 	get_protocol_type() const = 0;
 	virtual bool 		get_net_dev_val();
 	virtual uint32_t 	get_inline_sge_num() = 0;
-	virtual ibv_sge*	get_sge_lst_4_inline_send() = 0;
 	virtual ibv_sge*	get_sge_lst_4_not_inline_send() = 0;
+	ibv_sge*		get_sge_lst_4_inline_send() { return m_sge; };
```
Comment (liranoz12) [inline src/vma/proto/dst_entry.h]: Please update commit message with issue id

File: src/vma/proto/header.cpp [hunk 4]
```diff
@@ -39,25 +39,29 @@ void header::init()
 	memset(&m_header, 0, sizeof(m_header));
 	m_ip_header_len = 0;
 	m_transport_header_len = 0;
-	m_udp_header_len = 0;
 	m_total_hdr_len = 0;
 	m_aligned_l2_l3_len = 40;
+	m_is_valn_enabled = false;
```
Comment (liranoz12) [inline src/vma/proto/header.cpp]: should be is_vlan_enabled ?

File: src/vma/sock/sockinfo_tcp.cpp [hunk 5]
```diff
@@ -3443,6 +3443,17 @@ int sockinfo_tcp::setsockopt(int __level, int __optname,
 				val = *(int *)__optval;
 				val &= ~INET_ECN_MASK;
 				m_pcb.tos |= val & INET_ECN_MASK;
+				header_tos_updater du(m_pcb.tos);
+				update_header_field(&du);
+				// according to the linux
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_tcp.cpp]: Please add more detailed explanation

File: src/vma/sock/sockinfo_udp.cpp [hunk 6]
```diff
@@ -1216,6 +1218,16 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
 			case IP_TOS:
 				if (__optlen <= sizeof(int)) {
 					m_tos =(uint8_t) *(int *)__optval;
+					header_tos_updater du(m_pcp);
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_udp.cpp]: Can we move IP_TOS to sockinfo and avoid the duplication  ?

File: src/vma/sock/sockinfo_udp.cpp [hunk 7]
```diff
@@ -2817,18 +2829,18 @@ bool sockinfo_udp::prepare_to_close(bool process_shutdown) {
...
+	for (; dst_entry_iter != m_dst_entry_map.end(); dst_entry_iter++) {
 		if (!IN_MULTICAST_N(dst_entry_iter->second->get_dst_addr())) {
-			dst_entry_iter->second->set_ip_ttl(m_n_uc_ttl);
+			updater->update_field(*dst_entry_iter->second);
 		}
 	}
 
-	if (m_p_connected_dst_entry && !IN_MULTICAST_N(m_p_connected_dst_entry->get_dst_addr())) {
-		m_p_connected_dst_entry->set_ip_ttl(m_n_uc_ttl);
+	if (m_p_connected_dst_entry &&
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_udp.cpp]: Do we want to ignore it for MC ?

File: src/vma/proto/dst_entry.cpp (hunk 1 above)
Comment (rafiw) [reply src/vma/proto/dst_entry.cpp]: yes i doubled check and compared to the kernel

File: src/vma/proto/header.h (hunk 2 above)
Comment (rafiw) [reply src/vma/proto/header.h]: there isn't one

Comment (rafiw) [reply src/vma/proto/header.h]: there isn't one

File: src/vma/proto/header.cpp (hunk 4 above)
Comment (rafiw) [reply src/vma/proto/header.cpp]: done

File: src/vma/sock/sockinfo_tcp.cpp (hunk 5 above)
Comment (rafiw) [reply src/vma/sock/sockinfo_tcp.cpp]: added

File: src/vma/sock/sockinfo_udp.cpp (hunk 6 above)
Comment (rafiw) [reply src/vma/sock/sockinfo_udp.cpp]: there is a difference in behavior

**Comment by liranoz12** [comment]
Duplicated by #741

**Pull request closed without merging**